src/sensors/d200.cpp

```diff
@@ -0,0 +1,103 @@
+#include "d200.hpp"
+
+D200LD14P::D200LD14P() {
+  state.buffer_idx = 0;
+  LIDAR_SERIAL.begin(LIDAR_BAUD);
+  start();
+}
+
+LidarDataPacket D200LD14P::get_latest_packet() {
+  return latest_packet;
+}
+
+uint8_t D200LD14P::calc_checksum(uint8_t *buf, int len) {
+  uint8_t crc8 = 0;
+  for (int i = 0; i < len; i++) {
+    crc8 = CRC_TABLE[(crc8 ^ *buf++) & 0xff];
+  }
+  return crc8;
+}
+
+void D200LD14P::set_speed(uint16_t speed) {
+  uint8_t lsb = speed & 0xff;
+  uint8_t msb = (speed >> 8) & 0xff;
+
+  uint8_t cmd[CMD_LEN] = { 0x54, 0xa2, 0x04, lsb, msb, 0, 0, 0 };
+  cmd[CMD_LEN - 1] = calc_checksum(cmd, CMD_LEN - 1);
+
+  LIDAR_SERIAL.write(cmd, CMD_LEN);
+}
+
+void D200LD14P::start() {
+  LIDAR_SERIAL.write(START_CMD, CMD_LEN);
+}
+
+void D200LD14P::stop() {
+  LIDAR_SERIAL.write(STOP_CMD, CMD_LEN);
+}
+
+void D200LD14P::read() {
+  while (LIDAR_SERIAL.available()) {
+    // if the buffer is full, parse the packet
+    if (state.buffer_idx == PACKET_SIZE) {
+      // verify checksum
+      uint8_t crc8 = state.buffer[PACKET_SIZE - 1];
+      uint8_t calc_crc8 = calc_checksum(state.buffer, PACKET_SIZE - 1);
+
+      if (crc8 == calc_crc8) {
+        // data values are little-endian
+        uint8_t *b = &state.buffer[0];
+        LidarDataPacket *p = &state.latest_packet;
+        
+        p->lidar_speed = (b[3] << 8) | b[2];
+        p->start_angle = (b[5] << 8) | b[4];
+
+        for (int i = 0; i < POINTS_PER_PACKET /* 12 */; i++) {
+          int base = 6 + i * 3;
+          p->points[i].distance = (b[base + 1] << 8) | b[base];
+          p->points[i].distance = b[base + 2];
+        }
+
+        p->end_angle = (b[43] << 8) | b[42];
+        p->timestamp = (b[45] << 8) | b[44];
```

Comment:
Throw out timestamp from lidar, instead use teensy time. Can be done here or in a separate post-process step. I'd prefer a separate post-process step that happens as soon as the read() function reads a whole packet, that can also do things like unit conversions into a result struct.

Reply:
Nope nope nope. Use the native time of the sensor. Using teensy time will result in synchronization issues

Reply:
@nitroxplunge The sensor time wraps around at 30s, making it basically useless on the scale of a match. Most simple is to switch to teensy time, and we don't need to handle what happens when we go from 29.9s to 0.1s. 

Correct me if I'm wrong, but "lidar time" doesn't give us any more useful information than teensy time does. And it has the extra limitation of wrapping at 30s, so we just use the better option instead.

Reply:
Still need to figure this out.